### ISSUE TITLE
Tests refactoring

### ISSUE THREAD

**Issue by lukasz-szyszkowski** [ably/ably-asset-tracking-swift#185]
There are a couple of things that need to be refactored:

- creating objects got to be moved to helper class/method instead of keeping the reference to them as test class field
- boilerplate code should be moved to the shared container (the SPM Target would be the best solution)
- authentication tests got to be more accurate. Using `publisher.track` shouldn’t be enough to test that connection to `ably service` was a success. We need to introduce a `Subscriber` that could receive info about trackable from the `Publisher`

FYI: @KacperKluka

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/SDK-1222) by [Unito](https://www.unito.io)

**Comment by QuintinWillison** [comment]
Looks good. Centralising object creation and other 'test utilities' is never a bad idea, but please bear in mind the full lifecycle of these entities - including things like waiting for things to fully stop before concluding a test is finished. I can expand more on this if you need me to, but the biggest mistake I see in system testing generally - across the board - is that tests are allowed to overlap each other (i.e. the previous test had not fully completing executing / closing down things it had spawned).

Another tip with the tests, and I've not really looked into whether @KacperKluka does this much in the Android ones yet, is to inject a new UUID and/or current date/time into names used in the external systems (e.g. Ably service channel names). i.e. each test defines unique, ephemeral names for things like channels (/ tracking ids). This then means that even if tests do end up overlapping then there is a bit less chance that external variances will change behaviour from one entire-suite run to the next.

Finally. in terms of what we've been seeing elsewhere, please configure test runners to run tests serially. The 'optimisation' of running tests in parallel 'to speed up developer time' will often cause more issues - again relating to the use of a shared, external service.

**Comment by lukasz-szyszkowski** [comment]
- add semaphores mentioned https://github.com/ably/ably-asset-tracking-swift/pull/169#discussion_r707063547
- change HTTP method to POST for JWT creation https://github.com/ably/ably-asset-tracking-swift/pull/169#discussion_r708523091

**Comment by ben-xD** [comment]
I also had a few other comments about the naming in `extension TrackableState: StateSkippable {`: 

For example, `skippedLocationsAdd`, `skippedLocationsClear` and `skippedLocationsList` could read better / fix the singular/plurals: 
- `getSkippedLocations(for)`
- `addSkippedLocation`
- `clearSkippedLocations`

Or alternatively make the state related to skipped Locations, and just have `get`, `add` and `clear`. This sounds like a container already 😅

**Comment by lukasz-szyszkowski** [comment]
> I also had a few other comments about the naming in `extension TrackableState: StateSkippable`, e.g. [#182 (comment)](https://github.com/ably/ably-asset-tracking-swift/pull/182#commitcomment-56727912):
> 
> For example, `skippedLocationsAdd`, `skippedLocationsClear` and `skippedLocationsList` could read better / fix the singular/plurals:
> 
> * `getSkippedLocations(for)`
> * `addSkippedLocation`
> * `clearSkippedLocations`
> 
> Or alternatively make the state related to skipped Locations, and just have `get`, `add` and `clear`. This sounds like a container already 😅

This one needs to be a separate issue, it's not in the scope of this issue (it's about tests).

**Comment by lukasz-szyszkowski** [comment]
Since https://github.com/ably/ably-asset-tracking-swift/pull/219 and https://github.com/ably/ably-asset-tracking-swift/pull/220 are not merged yet, I've moved this issue into `on hold` column.
